src/main/scala/fi/vm/sade/hakurekisteri/integration/koski/KoskiDataHandler.scala

```diff
@@ -307,43 +277,40 @@ class KoskiDataHandler(suoritusRekisteri: ActorRef, arvosanaRekisteri: ActorRef,
-              logger.warn(s"Koski-suoritusarvosanojen ${s} tallennus henkilölle ${henkilöOid} epäonnistui.", e)
-              Future.successful(Left(e))
+              Future.successful(Left(new RuntimeException(s"Koski-suoritusarvosanojen $s tallennus henkilölle $henkilöOid epäonnistui.", e)))
           }
           case _ => Future.successful(Right(None))
         })
-
-      ).map{x =>
-        logger.info(s"Koski-suoritusten tallennus henkilölle ${henkilöOid} valmis.")
-        x
-      }
+      )
     })
   }
 
   def createSuorituksetJaArvosanatFromKoski(henkilo: KoskiHenkiloContainer): Seq[Seq[SuoritusArvosanat]] = {
-    val viimeisimmat = ensureAinoastaanViimeisinOpiskeluoikeusJokaisestaTyypista(henkilo.opiskeluoikeudet, henkilo.henkilö.oid)
-    if (henkilo.opiskeluoikeudet.size > viimeisimmat.size) {
-      logger.info("Filtteröitiin henkilöltä " + henkilo.henkilö.oid + " pois yksi tai useampia opiskeluoikeuksia. Ennen filtteröintiä: " + henkilo.opiskeluoikeudet.size + ", jälkeen: " + viimeisimmat.size)
-    }
-    suoritusArvosanaParser.getSuoritusArvosanatFromOpiskeluoikeudes(henkilo.henkilö.oid.getOrElse(""), viimeisimmat)
+    val henkiloOid = henkilo.henkilö.oid.get
+    suoritusArvosanaParser.getSuoritusArvosanatFromOpiskeluoikeudes(henkiloOid, halututOpiskeluoikeudetJaSuoritukset(henkiloOid, henkilo.opiskeluoikeudet))
+  }
+
+  private def suoritusDuplicates(suoritukset: Seq[SuoritusArvosanat]): Seq[Seq[SuoritusArvosanat]] = {
+    suoritukset.groupBy(_.suoritus.core).collect {
+      case (_, suoritusarvosanat) if suoritusarvosanat.size > 1 => suoritusarvosanat
+    }.toSeq
   }
 
   def processHenkilonTiedotKoskesta(koskihenkilöcontainer: KoskiHenkiloContainer,
                                     personOidsWithAliases: PersonOidsWithAliases,
                                     params: KoskiSuoritusHakuParams): Future[Seq[Either[Exception, Option[SuoritusArvosanat]]]] = {
-
-    koskihenkilöcontainer.henkilö.oid match {
-      case Some(henkilöOid) => {
-        val henkilonSuoritukset: Seq[SuoritusArvosanat] = createSuorituksetJaArvosanatFromKoski(koskihenkilöcontainer).flatten
-          .filter(s => henkilöOid.equals(s.suoritus.henkiloOid))
-
-        henkilonSuoritukset match {
-          case Nil => Future.successful(Seq(Right(None)))
-          case _ => overrideExistingSuorituksetWithNewSuorituksetFromKoski(henkilöOid, henkilonSuoritukset, personOidsWithAliases, params)
-        }
-      }
-      case None => Future.successful(Seq(Right(None)))
+    val henkiloOid = koskihenkilöcontainer.henkilö.oid.get
+    val suoritukset = createSuorituksetJaArvosanatFromKoski(koskihenkilöcontainer).flatten
+    val muidenSuoritukset = suoritukset.filter(_.suoritus.henkilo != henkiloOid)
+    if (muidenSuoritukset.nonEmpty) {
+      return Future.successful(Seq(Left(new RuntimeException(s"Henkilön $henkiloOid Koskitiedoista syntyi suorituksia muille henkilöille ${muidenSuoritukset.map(_.suoritus.henkilo).mkString(", ")}"))))
```

Comment:
Tää vaikuttaa jo vähän paranoidilta, vaikka kyllähän siinä tietty virhetilanteesta olisi kyse. Mut niin kauan kun pelataan noilla KoskiHenkiloContainereilla, joissa on tasan yksi henkilö sisällä, ei liene riskiä että näin kävisi.